src/aws_encryption_sdk/materials_managers/__init__.py

```diff
@@ -40,38 +49,285 @@ class EncryptionMaterialsRequest(object):
     :param dict encryption_context: Encryption context tied to `encrypted_data_keys`
-    :param bytes signing_key: Encoded signing key
+    :param bytes signing_key: Encoded signing key (optional)
+    :param keyring_trace: Any KeyRing trace entries (optional)
+    :type keyring_trace: list of :class:`KeyringTrace`
     """
 
-    algorithm = attr.ib(validator=attr.validators.instance_of(Algorithm))
-    data_encryption_key = attr.ib(validator=attr.validators.instance_of(DataKey))
-    encrypted_data_keys = attr.ib(validator=attr.validators.instance_of(set))
-    encryption_context = attr.ib(validator=attr.validators.instance_of(dict))
-    signing_key = attr.ib(default=None, validator=attr.validators.optional(attr.validators.instance_of(bytes)))
+    _encrypted_data_keys = attr.ib(
+        default=attr.Factory(list), validator=optional(deep_iterable(member_validator=instance_of(EncryptedDataKey)))
+    )
+    signing_key = attr.ib(default=None, repr=False, validator=optional(instance_of(bytes)))
+
+    def __init__(
+        self,
+        algorithm=None,
+        data_encryption_key=None,
+        encrypted_data_keys=None,
+        encryption_context=None,
+        signing_key=None,
+        **kwargs
+    ):  # noqa we define this in the class docstring
+        if algorithm is None:
+            raise TypeError("algorithm must not be None")
+
+        if encryption_context is None:
+            raise TypeError("encryption_context must not be None")
+
+        if data_encryption_key is None and encrypted_data_keys is not None:
+            raise TypeError("encrypted_data_keys cannot be provided without data_encryption_key")
+
+        if encrypted_data_keys is None:
+            encrypted_data_keys = []
+
+        super(EncryptionMaterials, self).__init__(
+            algorithm=algorithm,
+            encryption_context=encryption_context,
+            data_encryption_key=data_encryption_key,
+            **kwargs
+        )
+        self._setattr("signing_key", signing_key)
+        self._setattr("_encrypted_data_keys", encrypted_data_keys)
+        attr.validate(self)
+
+    @property
+    def encrypted_data_keys(self):
+        # type: () -> FrozenSet[EncryptedDataKey]
+        """Return a read-only version of the encrypted data keys.
+
+        :rtype: frozenset
+        """
+        return frozenset(self._encrypted_data_keys)
+
+    @property
+    def is_complete(self):
+        # type: () -> bool
+        """Determine whether these materials are sufficiently complete for use as decryption materials.
+
+        :rtype: bool
+        """
+        if self.data_encryption_key is None:
+            return False
+
+        if not self.encrypted_data_keys:
+            return False
+
+        if self.algorithm.signing_algorithm_info is not None and self.signing_key is None:
+            return False
+
+        return True
+
+    def add_data_encryption_key(self, data_encryption_key, keyring_trace):
```

Comment:
Re: https://github.com/aws/aws-encryption-sdk-python/pull/163#discussion_r299256972

I think a comment about _why_ the keyring trace is here?  Maybe it will resolve the other convo?